File: gibbs/models/ceos.py
import numpy as np
import attr

from gibbs.mixture import Mixture

R = 8.3144598


def check_bip(instance, attribute, value):
    if value.shape[0] != value.shape[1]:
        raise ValueError("BIP's must be a 2-dim symmetric array.")
    if value.shape[0] != len(instance.mixture.Tc):
        raise ValueError("BIP's have incompatible dimension with input data such as critical temperature.")


@attr.s
class CEOS(object):
    """
    docstring here
        :param object: 
    """
    mixture = attr.ib(type=Mixture)
    bip = attr.ib(type=np.ndarray, validator=[check_bip])
    _Z_c = attr.ib(type=float)
    _Omega_a = attr.ib(type=float)
    _Omega_b = attr.ib(type=float)
    _m = attr.ib(type=np.ndarray, default=None)
    _b_i = attr.ib(type=np.ndarray, default=None)

    def __attrs_post_init__(self):
        # Used like caches, to prevent several evaluations
        self._m = self.m
        self._b_i = self.b_i

    @property
    def n_components(self):
        return len(self.mixture.Pc)

    def Tr(self, T):
        if T < 0:
            raise ValueError('Temperature must be greater than zero.')
        return T / self.mixture.Tc

    def Pr(self, P):
        if P < 0:
            raise ValueError('Pressure must be greater than zero.')
        return P / self.mixture.Pc

    @property
    def m(self):
        return NotImplementedError('Abstract method')

    def a_i(self, T):
        return self._Omega_a * R ** 2.0 * self.mixture.Tc * self.mixture.Tc / self.mixture.Pc * self.alpha(T)

    @property
    def b_i(self):
        return self._Omega_b * R * self.mixture.Tc / self.mixture.Pc

    def alpha(self, T):
        return NotImplementedError('Abstract method')

    def A_i(self, P, T):
        return self.a_i(T) * P / ((R * T) ** 2.0)

    def B_i(self, P, T):
        return self._b_i * P / (R * T)

    def A_mix(self, P, T, z):
        return np.dot(z, np.dot(z, self.A_ij(P, T)))

    def B_mix(self, P, T, z):
        return np.dot(z, self.B_i(P, T))

    def A_ij(self, P, T):
        return (1 - self.bip) * np.sqrt(np.outer(self.A_i(P, T), self.A_i(P, T)))

    def calculate_Z_factor(self, P, T, z):
        return NotImplementedError('Abstract method')

    def calculate_Z_minimal_energy(self, P, T, z):
        Z_roots = self.calculate_Z_factor(P, T, z)
        previous_normalized_gibbs_energy = np.inf
        Z_min = None
        if len(Z_roots) == 1:
            Z_min = Z_roots[0]
        else:
            for Z in Z_roots:
                fugacity = self.calculate_fugacity(P, T, z, Z)
                ln_f = np.log(fugacity)
                current_normalized_gibbs = np.dot(z, ln_f)
                if current_normalized_gibbs <= previous_normalized_gibbs_energy:
                    Z_min = Z
        return Z_min

    def calculate_fugacity_coefficient(self, P, T, z, Z_factor):
        return NotImplementedError('Abstract method')

    def calculate_fugacity(self, P, T, z, Z_factor):
        return z * P * self.calculate_fugacity_coefficient(P, T, z, Z_factor)


@attr.s
class PengRobinson(CEOS):
    """
    docstring here
        :param CEOS: 
    """
    _Z_c = attr.ib(default=0.3074)
    _Omega_a = attr.ib(default=0.45724)
    _Omega_b = attr.ib(default=0.07780)

    @property
    def m(self):
        omega = self.mixture.omega
        m = 0.37464 + 1.54226 * omega - 0.26992 * omega * omega
        return m

    def alpha(self, T):
        return (1 + self._m * (1 - np.sqrt(self.Tr(T)))) * (1 + self._m * (1 - np.sqrt(self.Tr(T))))

    def calculate_Z_factor(self, P, T, z):
        A = self.A_mix(P, T, z)
        B = self.B_mix(P, T, z)

        # Obtaining the roots
        coeff0 = -(A * B - B ** 2.0 - B ** 3.0)
        coeff1 = A - 3 * B ** 2.0 - 2 * B
        coeff2 = -(1 - B)
        coeff3 = 1
        coefficients = np.array([coeff3, coeff2, coeff1, coeff0])
        Z_roots = np.roots(coefficients)

        # A threshold is applied in imaginary part, since it can be 
        # numerically spurious.
        Z_real_roots = Z_roots.real[np.abs(Z_roots.imag) < 1e-5]

        # Filtering non-physical roots
        Z_real_roots = Z_real_roots[Z_real_roots >= 0.0]

        if len(Z_real_roots) == 3:
            Z_real_roots = np.append(Z_real_roots[0], Z_real_roots[-1])

        return Z_real_roots

    def calculate_fugacity_coefficient(self, P, T, z, Z_factor):
        # Avoiding multiple function calls
        A = self.A_mix(P, T, z)
        B = self.B_mix(P, T, z)
        Bi_B = self.B_i(P, T) / B
        A_ij = self.A_ij(P, T)
        sqrted_2 = np.sqrt(2.)

        # Computing fugacity coefficients by parts
        first_term = Bi_B * (Z_factor - 1)
        second_term = np.log(Z_factor - B)
        third_term = A / (2 * sqrted_2 * B) * (Bi_B - 2 / A * np.dot(z, A_ij))
        fourth_term = np.log(
            (Z_factor + (1 + sqrted_2) * B) / (Z_factor + (1 - sqrted_2) * B)
        )
        ln_phi = first_term - second_term + third_term * fourth_term

        return np.exp(ln_phi)


@attr.s
class PengRobinson78(PengRobinson):
    """
    docstring here
        :param CEOS:
    """

    @property
    def m(self):
        omega = self.mixture.omega
        m_low = 0.37464 + 1.54226 * omega - 0.26992 * omega * omega
        m_high = 0.3796 + 1.485 * omega - 0.1644 * omega * omega \
            + 0.01667 * omega * omega * omega
        m_value = np.where(self.mixture.omega > 0.49, m_high, m_low)
        return m_value


@attr.s
class SoaveRedlichKwong(CEOS):
    """
    docstring here
        :param CEOS:
    """
    _Z_c = attr.ib(default=1/3)
    _Omega_a = attr.ib(default=0.42748)
    _Omega_b = attr.ib(default=0.08664)

    @property
    def m(self):
        omega = self.mixture.omega
        m_value = 0.480 + 1.574 * omega - 0.176 * omega * omega
        return m_value

    def alpha(self, T):
        return (1 + self._m * (1 - np.sqrt(self.Tr(T)))) * (1 + self._m * (1 - np.sqrt(self.Tr(T))))

    def calculate_Z_factor(self, P, T, z):
        A = self.A_mix(P, T, z)
        B = self.B_mix(P, T, z)

        # Obtaining the roots
        coeff0 = -A * B
        coeff1 = A - B - B ** 2.0
        coeff2 = -1
        coeff3 = 1
        coefficients = np.array([coeff3, coeff2, coeff1, coeff0])
        Z_roots = np.roots(coefficients)

        # A threshold is applied in imaginary part, since it can be
        # numerically spurious.
        Z_real_roots = Z_roots.real[np.abs(Z_roots.imag) < 1e-5]

        # Filtering non-physical roots
        Z_real_roots = Z_real_roots[Z_real_roots >= 0.0]

        if len(Z_real_roots) == 3:
            Z_real_roots = np.append(Z_real_roots[0], Z_real_roots[-1])

        return Z_real_roots

    def calculate_fugacity_coefficient(self, P, T, z, Z_factor):
        # Avoiding multiple function calls
        A = self.A_mix(P, T, z)
        B = self.B_mix(P, T, z)
        Bi_B = self.B_i(P, T) / B
        A_ij = self.A_ij(P, T)

        # Computing fugacity coefficients by parts
        first_term = Bi_B * (Z_factor - 1)
        second_term = np.log(Z_factor - B)
        third_term = A / B * (Bi_B - 2 / A * np.dot(z, A_ij))
        fourth_term = np.log(1 + B / Z_factor)
        ln_phi = first_term - second_term + third_term * fourth_term

        return np.exp(ln_phi)

File: gibbs/eos_cpp/ceos.cpp
#include <variant>
#include <optional>
#include <stdexcept>
#include <cmath>
#include <vector>
#include <Eigen/Dense>
#include "eos_cpp/mixture.hpp"
#include "eos_cpp/ceos.hpp"
#include "eos_cpp/constants.hpp"
#include "eos_cpp/solvers.hpp"

using namespace Eigen;
using namespace eos;
using namespace mixture;

/*
 * Abstract Cubic EoS member functions definitions.
 * */
namespace eos {

    Mixture CubicEos::get_mixture() const { return this->mixture; }

    ArrayXXd CubicEos::get_bip() const { return this->bip; }

    ArrayXd CubicEos::Tr(const double &T) const {
        if (T < 0) {
            throw std::invalid_argument("Temperature must be greater than zero.");
        }
        return T / this->mixture.Tc();
    }

    ArrayXd CubicEos::Pr(const double &P) const {
        if (P < 0) {
            throw std::invalid_argument("Pressure must be greater than zero.");
        }
        return P / this->mixture.Pc();
    }

    ArrayXd CubicEos::calculate_a(const double &T) const {
        ArrayXd Tc_squared = this->mixture.Tc().square();
        ArrayXd numerator = (this->_Omega_a * constants::R * constants::R * Tc_squared * this->alpha(T));
        ArrayXd denominator = this->mixture.Pc();
        return (numerator / denominator);
    }

    ArrayXd CubicEos::calculate_b() const {
        return (this->_Omega_b * constants::R * this->mixture.Tc() / this->mixture.Pc());
    }

    ArrayXd CubicEos::calculate_A(const double &P, const double &T) const {
        return (this->calculate_a(T) * P / (constants::R * T * constants::R * T));
    }

    ArrayXd CubicEos::calculate_B(const double &P, const double &T) const {
        return (this->calculate_b() * P / (constants::R * T));
    }

    ArrayXXd CubicEos::calculate_A_ij(const double &P, const double &T) const {
        ArrayXd A = this->calculate_A(P, T);
        ArrayXXd bips = this->bip;
        auto size_A = A.size();
        ArrayXXd A_ij(size_A, size_A);

        for (int i = 0; i < size_A; ++i)
            for (int j = 0; j < size_A; ++j)
                A_ij(i, j) = (1 - bips(i, j)) * std::sqrt(A(i) * A(j));

        return A_ij;
    }

    double CubicEos::calculate_A_mix(const double &P, const double &T, const ArrayXd &z) const {
        VectorXd z_matrix = z.matrix();
        MatrixXd A_ij = this->calculate_A_ij(P, T).matrix();
        VectorXd z_dot_A_ij = (z_matrix.transpose() * A_ij);
        return z_matrix.dot(z_dot_A_ij);
    }

    double CubicEos::calculate_B_mix(const double &P, const double &T, const ArrayXd &z) const {
        VectorXd z_matrix = z.matrix();
        VectorXd B_i = this->calculate_B(P, T).matrix();
        return z_matrix.dot(B_i);
    }

    ArrayXd CubicEos::calculate_Z_factor(const double &P, const double &T, const ArrayXd &z) const {
        Vector4d coefficients = this->calculate_Z_cubic_polynomial_coeffs(P, T, z);

        auto result_roots = eos::solvers::cubic_polynomial_real_positive_roots(coefficients);

        Array2d empty_array;
        if (std::holds_alternative<ArrayXd>(result_roots)) {
            auto Z_roots_full = std::get<ArrayXd>(result_roots);
            if (Z_roots_full.size() < 1)
                return empty_array;

            auto smallest_root = Z_roots_full.minCoeff();
            auto greatest_root = Z_roots_full.maxCoeff();
            auto Z_positive_roots = eos::solvers::select_polynomial_real_positive_roots(
                    smallest_root,
                    greatest_root);

            if (std::holds_alternative<ArrayXd>(Z_positive_roots)) {
                return std::get<ArrayXd>(Z_positive_roots);
            }

            return empty_array;

        }

        return empty_array;

    }

    std::variant<double, std::nullopt_t> CubicEos::calculate_Z_minimal_energy(const double &P, const double &T, const ArrayXd &z) const {
        ArrayXd Z_roots = this->calculate_Z_factor(P, T, z);
        auto previous_normalized_gibbs_energy = std::numeric_limits<double>::infinity();
        std::variant<double, std::nullopt_t> Z_min;
        Z_min = std::nullopt;

        if (Z_roots.size() == 1)
            return Z_roots(0);
        else {
            for (int i = 0; i < Z_roots.size(); ++i) {
                ArrayXd fugacity = this->calculate_fugacity(P, T, z, Z_roots(i));
                ArrayXd ln_f = fugacity.log();
                double current_normalized_gibbs = z.matrix().dot(ln_f.matrix());
                if (current_normalized_gibbs <= previous_normalized_gibbs_energy)
                    Z_min = Z_roots(i);
            }
        }
        return Z_min;
    }

    ArrayXd CubicEos::calculate_fugacity(const double &P, const double &T, const ArrayXd &z, const double &Z_factor) const {
        return z * P * this->calculate_fugacity_coefficients(P, T, z, Z_factor);
    }

}

/*
 * Peng-Robinson member functions definitions.
 * */
namespace eos {

    ArrayXd PengRobinson::m() const {
        ArrayXd omega = this->mixture.omega();
        return (0.37464 + 1.54226 * omega - 0.26992 * omega * omega);
    }

    ArrayXd PengRobinson::alpha(const double &T) const {
        return ((1 + this->m() * (1 - this->Tr(T).sqrt())) * (1 + this->m() * (1 - this->Tr(T).sqrt())));
    }

    Vector4d PengRobinson::calculate_Z_cubic_polynomial_coeffs(const double &P, const double &T, const ArrayXd &z)
    const {
        double A = this->calculate_A_mix(P, T, z);
        double B = this->calculate_B_mix(P, T, z);

        // Obtaining the roots
        Vector4d coefficients;
        coefficients(0) = - (A * B - B * B - B * B * B);
        coefficients(1) = A - 3 * B * B - 2 * B;
        coefficients(2) = B - 1;
        coefficients(3) = 1;

        return coefficients;

    }

    ArrayXd PengRobinson::calculate_fugacity_coefficients(const double &P, const double &T, const ArrayXd &z,
                                                          const double &Z_factor) const {
        // Avoiding multiple function calls
        double A = this->calculate_A_mix(P, T, z);
        double B = this->calculate_B_mix(P, T, z);
        ArrayXd Bi_B = this->calculate_B(P, T) / B;
        ArrayXXd A_ij = this->calculate_A_ij(P, T);
        double sqrted_2 = std::sqrt(double{2.0});

        auto N = this->n_components();
        ArrayXd ln_phi(N);
        for (int i = 0; i < N; ++i) {
            // Computing fugacity coefficients by parts
            auto first_term = (Bi_B(i) * (Z_factor - 1.0));
            auto second_term = std::log(Z_factor - B);
            auto z_dot_A_ij = 0.0;
            for (int j = 0; j < N; ++j)
                z_dot_A_ij += z(j) * A_ij(i, j);
            auto third_term = A / (2.0 * sqrted_2 * B) * (Bi_B(i) - 2.0 / A * z_dot_A_ij);
            auto fourth_term = std::log(
                    (Z_factor + (1.0 + sqrted_2) * B) / (Z_factor + (1.0 - sqrted_2) * B)
                    );
            ln_phi(i) = first_term - second_term + third_term * fourth_term;
        }

        return ln_phi.exp();

    }
}

/*
 * Peng-Robinson 78 member functions definitions.
 * */
namespace eos {

    ArrayXd PengRobinson78::m() const {
        auto omega = this->mixture.omega();
        std::vector<double> m_vector;
        for (int i = 0; i < omega.size(); ++i) {
            double m_value;
            if (omega(i) <= 0.49)
                m_value = 0.37464 + 1.54226 * omega(i) - 0.26992 * omega(i) * omega(i);
            else
                m_value = 0.3796 + 1.485 * omega(i) - 0.1644 * omega(i) * omega(i) + 0.01667 * omega(i) * omega(i) *
                        omega(i);
            m_vector.push_back(m_value);
        }

        return Map<ArrayXd>(m_vector.data(), m_vector.size());
    }
}